### PR TITLE
Add user_group_ids field to elasticache replication group

### LINKED ISSUE #20328: Add user_group argument to aws_elasticache_replication_group
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Recently user and user group resources for elasticache were added to the terraform provider. However there is not currently a way to associate a user group with a replication group.

### New or Affected Resource(s)

Impacted Existing Data Sources

- aws_elasticache_replication_group: will have a new list of user_group_ids to associate user groups with the relevant replication group 

### Potential Terraform Configuration

```
resource "aws_elasticache_replication_group" "test" {
  replication_group_id          = test_replication_group
  replication_group_description = "test description"
  node_type                     = "cache.t3.small"
  port                          = 6379
  apply_immediately             = true
  auto_minor_version_upgrade    = false
  maintenance_window            = "tue:06:30-tue:07:30"
  snapshot_window               = "01:00-02:00"
  user_group_ids                = [aws_elasticache_user_group.test.id]
}

resource "aws_elasticache_user" "test" {
  user_id       = test_user_group
  user_name     = "default"
  access_string = "on ~app::* -@all +@read +@hash +@bitmap +@geo -setbit -bitfield -hset -hsetnx -hmset -hincrby -hincrbyfloat -hdel -bitop -geoadd -georadius -georadiusbymember"
  engine        = "REDIS"
  passwords     = ["password123456789"]
}
resource "aws_elasticache_user_group" "test" {
  user_group_id = test_user_group
  engine        = "REDIS"
  user_ids      = [aws_elasticache_user.test.user_id]
}
```

### References

https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/Clusters.RBAC.html
https://docs.aws.amazon.com/AmazonElastiCache/latest/APIReference/API_CreateReplicationGroup.html#API_CreateReplicationGroup_RequestParameters

### PR DESCRIPTION
**Pull request by jamesglennan**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #20328

Output from acceptance testing:

```
$ make test TESTARGS='-run=TestAccAWSElasticacheReplicationGroup'
ok      github.com/terraform-providers/terraform-provider-aws/aws       1.884s
```
```
$ make testacc TESTARGS='-run=TestAccAWSElasticacheReplicationGroup'
```

### CODE REVIEW

**Comment by jamesglennan** [comment]
When I put this PR together, I had thought that a set of ids could be accepted, but upon testing it turns out that it accepts an array of strings, but will only accept one.

Should I leave this as it is `user_grou_ids= [test]` , or should i modify it to something more like `user_group_id = 'test'`

I'm thinking AWS would have made is this way for a reason, perhaps to support it in the future?

File: aws/resource_aws_elasticache_replication_group.go [hunk 1]
```diff
@@ -276,6 +276,15 @@ func resourceAwsElasticacheReplicationGroup() *schema.Resource {
...
 			},
+			"user_group_ids": {
+				Type:       schema.TypeSet,
+				ConfigMode: 0,
+				Optional:   true,
+				Default:    nil,
+				Elem:       &schema.Schema{Type: schema.TypeString},
+				MaxItems:   1, //at the moment the aws sdk only supports 1 user group id to be associated with a cluster
+				Set:        schema.HashString,
+			},
```
Comment (ramekin) [inline aws/resource_aws_elasticache_replication_group.go:287]: I am pretty sure this conflicts with the `auth_token` argument, so I think you should add `ConflictsWith` (here and then on the `auth_token` argument).

Like this:
```suggestion
			"user_group_ids": {
				Type:          schema.TypeSet,
				ConfigMode:    0,
				Optional:      true,
				Default:       nil,
				Elem:          &schema.Schema{Type: schema.TypeString},
				MaxItems:      1, //at the moment the aws sdk only supports 1 user group id to be associated with a cluster
				Set:           schema.HashString,
				ConflictsWith: []string{"auth_token"},
			},
```

Comment (jamesglennan) [reply aws/resource_aws_elasticache_replication_group.go]: Ah thankyou, yes, i've also added the field to the auth_token block as well

**Comment by hpdobrica** [comment]
Hey @jamesglennan, I needed multiple user groups provided to the replication group,  and am currently using null resource for this, so i came at the same issue as you did.

Take a look at this, it seems that taking only single-element array as input is an error within documentation, rather than error with the implementation:
https://github.com/aws/aws-cli/issues/6178

**Comment by jamesglennan** [comment]
@hpdobrica That issue looks related to multiple users to a user group, not multiple user groups to a replication group as far as i can tell?

**Comment by hpdobrica** [comment]
oh sorry, my bad

**Comment by renan** [comment]
Yes, please! This would be a good addition now since we can define groups and users would be nice to associate it with the replication group itself!

**Comment by kmai** [comment]
Any news on this?

**Comment by ricardosouzamorais** [comment]
Hello guys,

When are you guys planning to release this version ?

**Comment by daniel-laszlo** [comment]
Bump, I would also need this.

**Comment by jamesglennan** [comment]
@drarnold @zhelding I have updated my code and i created a new test for this function- not sure what the process from here is?

**Comment by stevehipwell** [comment]
@jamesglennan great work, this area has been neglected for quite a while and this change will be very useful.

On a slight tangent, have you considered the workflow to add a user group to an existing replication group? This is achievable with the `aws elasticache modify-replication-group`. There is also the flow to add a user to an existing user group via `aws elasticache modify-user-group`. All of these flows are currently only available via null resources to call the commands directly.

**Comment by drarnold** [comment]
@jamesglennan I am in the same boat because I don't know the next step, but I figure we can tag the reviewer for assistance.

@ramekin would we be able to get this reviewed?

**Comment by zhelding** [comment]
Hi @jamesglennan. No further action is required on your end for merging this pull request.

This pull request will now be addressed via our standard prioritization process. Unfortunately, given the large quantity of issues and pull requests we receive, we are not always able to review every pull request in a timely manner.

We prioritize contributions that provide the greatest value to the largest number of users -- which we primarily infer based on the number of :thumbsup: [reactions](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/).  Additionally, an explanation of the particular use case for a contribution can provide information on how the changes might be useful in a broader context.

You can get a view of the wider development plans for the provider via our [road map](https://github.com/hashicorp/terraform-provider-aws/blob/main/ROADMAP.md), published quarterly.

Thank you again for your contribution and your patience.

**Comment by karpenkoKhAI** [comment]
Hi everyone, do we have any updates on it?

**Comment by drarnold** [comment]
@jamesglennan 
Just wanted to follow up on this. I think we are also going to need to get the changelog file update added.
[Here](https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/pullrequest-submission-and-lifecycle.md#changelog-process) are the steps on how to add this update for the PR.

I hope that this helps!

**Comment by karpenkoKhAI** [comment]
Hey, do you have any timelines about implementing this feature? It is really needed feature

**Comment by jamesglennan** [comment]
I'm waiting from reviews from the TF staff- cant really do anything else until they approve

**Comment by nari1021** [comment]
Plz, release to this ASAP...!!

**Comment by nari1021** [comment]
@ewbankkit @jamesglennan @breathingdust  Could you give us a piece of advice about this issue?

**Comment by jamesglennan** [comment]
@zhelding Is there any chance i can get a review on this, seems like a lot of folks are keen to get this in? Thanks in advance

**Pull request merged**